File: web/src/entities/bossSaw.ts
import { Direction } from '../constants.js';
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';

export class BossSaw extends Entity {
  private readonly lifeTime = 60 * 6;
  private frameCounter = 0;
  private readonly saw: Animation;
  private readonly speed: ImmutableVec2;

  constructor(direction: Direction) {
    super();
    this.speed = direction === Direction.LEFT
      ? new ImmutableVec2(-2, 0)
      : new ImmutableVec2(2, 0);
    Sound.playSample('data/sounds/boss_saw');
    this.saw = Resource.getAnimation('data/images/saw.bmp');
    this.setSize(new ImmutableVec2(this.saw.getFrameWidth(), this.saw.getFrameHeight()));
  }

  getLayer(): number { return 3; }

  override onRespawn(): void { this.remove(); }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getDrawPositionX() + offsetX - this.getHalfSize().x);
    const y = Math.round(this.getDrawPositionY() + offsetY - this.getHalfSize().y);
    this.saw.drawFrame(ctx, Math.floor(this.frameCounter / 10), x, y);
  }

  override update(): void {
    this.frameCounter++;
    if (this.frameCounter > this.lifeTime) this.remove();
    this.setPosition(new ImmutableVec2(this.getPosition()).add(this.speed));
    if (this.myRoom.getHero().Collides(this)) this.myRoom.getHero().kill();
  }
}

File: wii/src/entities/boss_saw.cpp
#include "entities/boss_saw.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "room.h"
#include <cmath>

BossSaw::BossSaw(Direction direction) {
    speed_ = (direction == Direction::LEFT) ? Vec2(-2, 0) : Vec2(2, 0);
    Sound::playSample("data/sounds/boss_saw");
    saw_ = Resource::getAnimation("data/images/saw.bmp");
    setSize(Vec2(static_cast<float>(saw_->getFrameWidth()),
                 static_cast<float>(saw_->getFrameHeight())));
}

void BossSaw::onRespawn() {
    remove();
}

void BossSaw::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(getDrawPositionX() + offsetX - halfSize_.x));
    int y = static_cast<int>(std::round(getDrawPositionY() + offsetY - halfSize_.y));
    saw_->drawFrame(renderer, sawFrameCounter_ / 10, x, y);
}

void BossSaw::update() {
    sawFrameCounter_++;
    if (sawFrameCounter_ > lifeTime_) remove();
    position_ += speed_;
    if (room_->getHero()->Collides(*this)) room_->getHero()->kill();
}
